File: admin/src/types/goal.ts
import { Timestamp } from 'firebase/firestore';

/**
 * Tipos de metas
 */
export type GoalPeriod = 'weekly' | 'monthly';
export type GoalTargetType = 'users' | 'kiosks' | 'league' | 'all';

/**
 * Métricas de una meta
 */
export interface GoalMetrics {
  llamadas: number;      // Meta de llamadas/deals creados
  colocacion: number;    // Meta de colocación en pesos (amount)
}

/**
 * Modelo de Meta Comercial
 */
export interface Goal {
  id: string;
  name: string;                    // Nombre descriptivo (ej: "Meta Semanal - Enero 2025")
  period: GoalPeriod;              // 'weekly' o 'monthly'
  targetType: GoalTargetType;      // 'users', 'kiosks', 'league', o 'all'
  targetIds: string[];             // Array de IDs de usuarios, kioscos, o liga
  targetNames: string[];           // Array de nombres (para UI)
  metrics: GoalMetrics;            // Metas numéricas
  startDate: Timestamp;            // Fecha de inicio
  endDate: Timestamp;              // Fecha de fin
  active: boolean;                 // Si está activa
  createdAt: Timestamp;
  updatedAt: Timestamp;
  createdBy: string;               // User ID que creó la meta
}

/**
 * DTO para crear/actualizar meta
 */
export interface GoalFormData {
  name: string;
  period: GoalPeriod;
  targetType: GoalTargetType;
  targetIds: string[];           // Múltiples IDs
  targetNames: string[];         // Múltiples nombres
  metrics: {
    llamadas: string;    // String para input form
    colocacion: string;  // String para input form
  };
  startDate: string;     // ISO date string
  endDate: string;       // ISO date string
  active: boolean;
}

/**
 * Progreso de una meta
 */
export interface GoalProgress {
  goalId: string;
  userId?: string;
  userName?: string;
  current: {
    llamadas: number;
    colocacion: number;
  };
  target: GoalMetrics;
  percentage: {
    llamadas: number;
    colocacion: number;
  };
  onTrack: boolean;
  updatedAt: Timestamp;
}

File: app/src/main/java/com/promotoresavivatunegocio_1/models/Goal.kt
package com.promotoresavivatunegocio_1.models

import com.google.firebase.Timestamp
import com.google.gson.annotations.SerializedName

/**
 * Modelo de Meta Comercial
 */
data class Goal(
    @SerializedName("id")
    val id: String = "",

    @SerializedName("name")
    val name: String = "",

    @SerializedName("period")
    val period: GoalPeriod = GoalPeriod.WEEKLY,

    @SerializedName("targetType")
    val targetType: GoalTargetType = GoalTargetType.ALL,

    @SerializedName("targetId")
    val targetId: String? = null,

    @SerializedName("targetName")
    val targetName: String? = null,

    @SerializedName("metrics")
    val metrics: GoalMetrics = GoalMetrics(),

    @SerializedName("startDate")
    val startDate: Timestamp? = null,

    @SerializedName("endDate")
    val endDate: Timestamp? = null,

    @SerializedName("active")
    val active: Boolean = false,

    @SerializedName("createdAt")
    val createdAt: Timestamp? = null,

    @SerializedName("updatedAt")
    val updatedAt: Timestamp? = null,

    @SerializedName("createdBy")
    val createdBy: String = ""
)

/**
 * Tipo de periodo de meta
 */
enum class GoalPeriod {
    @SerializedName("weekly")
    WEEKLY,

    @SerializedName("monthly")
    MONTHLY
}

/**
 * Tipo de objetivo de meta
 */
enum class GoalTargetType {
    @SerializedName("kiosk")
    KIOSK,

    @SerializedName("seller")
    SELLER,

    @SerializedName("all")
    ALL
}

/**
 * Métricas de una meta
 */
data class GoalMetrics(
    @SerializedName("llamadas")
    val llamadas: Int = 0,

    @SerializedName("colocacion")
    val colocacion: Double = 0.0
)

/**
 * Progreso de una meta
 */
data class GoalProgress(
    @SerializedName("goalId")
    val goalId: String = "",

    @SerializedName("userId")
    val userId: String? = null,

    @SerializedName("userName")
    val userName: String? = null,

    @SerializedName("current")
    val current: GoalMetrics = GoalMetrics(),

    @SerializedName("target")
    val target: GoalMetrics = GoalMetrics(),

    @SerializedName("percentage")
    val percentage: PercentageMetrics = PercentageMetrics(),

    @SerializedName("onTrack")
    val onTrack: Boolean = false,

    @SerializedName("updatedAt")
    val updatedAt: Timestamp? = null
)

/**
 * Porcentajes de cumplimiento
 */
data class PercentageMetrics(
    @SerializedName("llamadas")
    val llamadas: Int = 0,

    @SerializedName("colocacion")
    val colocacion: Int = 0
)
